lib/ruby_lsp/requests/on_type_formatting.rb

```diff
@@ -0,0 +1,132 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # ![On type formatting demo](../../misc/on_type_formatting.gif)
+    #
+    # The [on type formatting](https://microsoft.github.io/language-server-protocol/specification#textDocument_onTypeFormatting)
+    # request formats code as the user is typing. For example, automatically adding `end` to class definitions.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Foo # <-- upon adding a line break, on type formatting is triggered
+    #   # <-- cursor ends up here
+    # end # <-- end is automatically added
+    # ```
+    class OnTypeFormatting < BaseRequest
+      extend T::Sig
+
+      END_REGEXES = T.let([
+        /(if|unless|for|while|class|module|until|def|case).*/,
+        /.*\sdo/,
+      ], T::Array[Regexp])
+
+      sig { params(document: Document, position: Document::PositionShape, trigger_character: String).void }
+      def initialize(document, position, trigger_character)
+        super(document)
+
+        scanner = Document::Scanner.new(document.source)
+        line_begin = position[:line] == 0 ? 0 : scanner.find_position({ line: position[:line] - 1, character: 0 })
```

Comment:
Not that I know of. I'm open to whatever the team prefers.

Reply:
In terms of type checking, both options would provide a similar level of guarantees, since Sorbet knows `position[:line]` is an Integer and thus knows what `==` should be. It also type checks that `position` matches all that is expected.

It would be a style choice, but since we have no standards, I'd say we can use whatever we prefer.

Reply:
Yup that's fine with me. Thanks for the info!